File: kafka_dev/src/main/java/com/wrp/partitioner/MyPartitioner.java
package com.wrp.partitioner;

import org.apache.kafka.clients.producer.Partitioner;
import org.apache.kafka.common.Cluster;

import java.util.Map;

/**
 * @ClassName MyPartitioner
 * @Author LYleonard
 * @Date 2020/5/24 0:31
 * @Description 自定义kafka的分区函数
 * Version 1.0
 **/
public class MyPartitioner implements Partitioner {
    /**
     * Compute the partition for the given record.
     * 通过这个方法来实现消息要去哪一个分区中
     *
     * @param topic      The topic name
     * @param key        The key to partition on (or null if no key)
     * @param keyBytes   The serialized key to partition on( or null if no key)
     * @param value      The value to partition on or null
     * @param valueBytes The serialized value to partition on or null
     * @param cluster    The current cluster metadata
     */
    public int partition(String topic, Object key, byte[] keyBytes, Object value, byte[] valueBytes, Cluster cluster) {
        // 获取topic分区数
        int partitons = cluster.partitionCountForTopic(topic);
        //key.hashCode()可能会出现负数 -1 -2 0 1 2
        //Math.abs 取绝对值
        return Math.abs(key.hashCode() % partitons);
    }

    /**
     * This is called when partitioner is closed.
     */
    public void close() {
    }

    /**
     * Configure this class with the given key-value pairs
     *
     * @param configs
     */
    public void configure(Map<String, ?> configs) {
    }
}

File: flink_dev/src/main/scala/io/wrp/operators/MyPartitioner.scala
package io.wrp.operators

import org.apache.flink.api.common.functions.Partitioner

/**
  * @author LYleonard
  * @date 2020/7/6 22:18
  * @description 自定义分区策略，实现不同分区的数据发送到不同分区中处理
  */
class MyPartitioner extends Partitioner[String]{
  override def partition(key: String, numPartitions: Int): Int = {
    println("分区个数为：" + numPartitions)
    if(key.contains("hello")){
      0
    } else {
      1
    }
  }
}
